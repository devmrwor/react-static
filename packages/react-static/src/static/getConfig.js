/* eslint-disable import/no-dynamic-require */

import React from 'react'
import nodePath from 'path'
import chokidar from 'chokidar'
import resolveFrom from 'resolve-from'
import fs from 'fs-extra'

import getDirname from '../utils/getDirname'
import {
  cleanSlashes,
  cutPathToRoot,
  isAbsoluteUrl,
  makeHookReducer,
} from '../utils'

// the default static.config.js location
const DEFAULT_NAME_FOR_STATIC_CONFIG_FILE = 'static.config.js'
const DEFAULT_PATH_FOR_STATIC_CONFIG = nodePath.resolve(
  nodePath.join(process.cwd(), DEFAULT_NAME_FOR_STATIC_CONFIG_FILE)
)
const DEFAULT_ROUTES = [{ path: '/' }]
const DEFAULT_ENTRY = 'index.js'

export const buildConfig = async (config = {}) => {
  // path defaults
  config.paths = {
    root: nodePath.resolve(process.cwd()),
    src: 'src',
    dist: 'dist',
    temp: 'tmp',
    devDist: 'tmp/dev-server',
    public: 'public',
    plugins: 'plugins',
    pages: 'src/pages',
    nodeModules: 'node_modules',
    assets: '',
    ...(config.paths || {}),
  }

  // Use the root to resolve all other relative paths
  const resolvePath = relativePath =>
    nodePath.resolve(config.paths.root, relativePath)

  // Resolve all paths
  const DIST =
    process.env.REACT_STATIC_ENV === 'development'
      ? resolvePath(config.paths.devDist || config.paths.dist)
      : resolvePath(config.paths.dist)

  const ASSETS = nodePath.resolve(DIST, config.paths.assets)

  const paths = {
    ROOT: config.paths.root,
    LOCAL_NODE_MODULES: nodePath.resolve(getDirname(), '../../node_modules'),
    SRC: resolvePath(config.paths.src),
    PAGES: resolvePath(config.paths.pages),
    DIST,
    ASSETS,
    PLUGINS: resolvePath(config.paths.plugins),
    TEMP: resolvePath(config.paths.temp),
    PUBLIC: resolvePath(config.paths.public),
    NODE_MODULES: resolvePath(config.paths.nodeModules),
    EXCLUDE_MODULES:
      config.paths.excludeResolvedModules ||
      resolvePath(config.paths.nodeModules),
    PACKAGE: resolvePath('package.json'),
    HTML_TEMPLATE: nodePath.join(DIST, 'index.html'),
    STATIC_DATA: nodePath.join(ASSETS, 'staticData'),
  }

  let siteRoot = ''
  let basePath = ''

  if (process.env.REACT_STATIC_ENV === 'development') {
    basePath = cleanSlashes(config.devBasePath)
  } else if (process.env.REACT_STATIC_STAGING === 'true') {
    siteRoot = cutPathToRoot(config.stagingSiteRoot, '$1')
    basePath = cleanSlashes(config.stagingBasePath)
  } else {
    siteRoot = cutPathToRoot(config.siteRoot, '$1')
    basePath = cleanSlashes(config.basePath)
  }

  const publicPath = `${cleanSlashes(`${siteRoot}/${basePath}`)}/`

  let assetsPath = cleanSlashes(config.assetsPath || paths.assets)
  if (assetsPath && !isAbsoluteUrl(assetsPath)) {
    assetsPath = `/${cleanSlashes(`${basePath}/${assetsPath}`)}/`
  }

  // Add the project root as a plugin. This allows the dev
  // to use the plugin api directory in their project if they want
  const plugins = [paths.ROOT, ...(config.plugins || [])]

  // Defaults
  config = {
    // Defaults
    entry: nodePath.join(paths.SRC, DEFAULT_ENTRY),
    getSiteData: () => ({}),
    renderToElement: Comp => <Comp />,
    renderToHtml: (render, comp) => render(comp),
    prefetchRate: 3,
    maxThreads: Infinity,
    disableRouteInfoWarning: false,
    disableRoutePrefixing: false,
    outputFileRate: 100,
    extensions: ['.js', '.jsx'], // TODO: document
    getRoutes: async () => DEFAULT_ROUTES,
    // Config Overrides
    ...config,
    // Materialized Overrides
    plugins,
    paths,
    siteRoot,
    basePath,
    publicPath,
    assetsPath,
    extractCssChunks: config.extractCssChunks || false,
    inlineCss: config.inlineCss || false,
  }

  // Set env variables to be used client side
  process.env.REACT_STATIC_PREFETCH_RATE = config.prefetchRate
  process.env.REACT_STATIC_DISABLE_ROUTE_INFO_WARNING =
    config.disableRouteInfoWarning
  process.env.REACT_STATIC_DISABLE_ROUTE_PREFIXING =
    config.disableRoutePrefixing

  const resolvePlugin = location => {
    let options = {}
    if (Array.isArray(location)) {
      location = location[0]
      options = location[1] || {}
    }

    let getHooks = () => ({})

    const originalLocation = location

    if (!fs.pathExistsSync(originalLocation)) {
      // If not an absolute path try from the plugins directory
      location = nodePath.resolve(paths.PLUGINS, originalLocation)
      if (!fs.pathExistsSync(location)) {
        // If not in the plugins directory, try from the currentworking directory
        try {
          location = resolveFrom(process.cwd(), originalLocation)
          location = location.includes('.')
            ? nodePath.resolve(location, '../')
            : location
        } catch (err) {
          //
        }
        if (!fs.pathExistsSync(location)) {
          location = null
        }
      }
    }

    // TODO: We have to do this because we don't have a good mock for process.cwd() :(
    if (process.env.NODE_ENV !== 'test' && !location) {
      throw new Error(
        `Oh crap! Could not find a plugin directory for the plugin: "${originalLocation}". We must bail!`
      )
    }

    let nodeLocation = nodePath.join(location, 'node.api.js')
    let browserLocation = nodePath.join(location, 'browser.api.js')

    try {
      // Get the hooks for the node api
      if (fs.pathExistsSync(nodeLocation)) {
        getHooks = require(nodeLocation).default
      } else {
        nodeLocation = null
      }

      // Detect if the browser plugin entry exists, and provide the nodeResolver to it
      browserLocation = fs.pathExistsSync(browserLocation)
        ? browserLocation
        : null

      const resolvedPlugin = {
        location,
        nodeLocation,
        browserLocation,
        options,
        hooks: getHooks(options) || {},
      }

      // Recursively resolve plugins
      if (resolvedPlugin.plugins) {
        resolvedPlugin.plugins = resolvedPlugin.plugins.map(resolvePlugin)
      }

      return resolvedPlugin
    } catch (err) {
      console.error(
        `The following error occurred in the plugin located at "${nodeLocation}"`
      )
      throw err
    }
  }

  config.plugins = config.plugins.map(resolvePlugin)

  const configHook = makeHookReducer(config.plugins, 'config')
  config = await configHook(config)

  return config
}

const buildConfigFromPath = async configPath => {
  const filename = nodePath.resolve(configPath)
  delete require.cache[filename]
  try {
    const config = require(filename).default
    return buildConfig(config)
  } catch (err) {
    console.error(err)
    return {}
  }
}

// Retrieves the static.config.js from the current project directory
export default (async function getConfig(
  configPath = DEFAULT_PATH_FOR_STATIC_CONFIG,
  subscription
) {
  const config = await buildConfigFromPath(configPath)

  if (subscription) {
    // If subscribing, return a never ending promise
    return new Promise(() => {
      chokidar.watch(configPath).on('all', async () => {
        subscription(await buildConfigFromPath(configPath))
      })
    })
  }

  return config
})
