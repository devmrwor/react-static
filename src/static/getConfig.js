/* eslint-disable import/no-dynamic-require */

import React from 'react'
import path from 'path'
import chokidar from 'chokidar'

import { pathJoin } from '../utils/shared'
import { reloadRoutes } from './webpack'
import getDirname from '../utils/getDirname'

const REGEX_TO_CUT_TO_ROOT = /(\..+?)\/.*/g
const REGEX_TO_REMOVE_TRAILING_SLASH = /^\/{0,}/g
const REGEX_TO_REMOVE_LEADING_SLASH = /\/{0,}$/g

const DEFAULT_NAME_FOR_STATIC_CONFIG_FILE = 'static.config.js'
// the default static.config.js location
const DEFAULT_PATH_FOR_STATIC_CONFIG = path.resolve(
  path.join(process.cwd(), DEFAULT_NAME_FOR_STATIC_CONFIG_FILE)
)
const DEFAULT_ROUTES = [{ path: '/' }]
const DEFAULT_ENTRY = 'index.js'
const PATH_404 = '404'

export const cutPathToRoot = (string = '') => string.replace(REGEX_TO_CUT_TO_ROOT, '$1')

export const trimLeadingAndTrailingSlashes = (string = '') =>
  string.replace(REGEX_TO_REMOVE_TRAILING_SLASH, '').replace(REGEX_TO_REMOVE_LEADING_SLASH, '')

export const throwErrorIfRouteIsMissingPath = route => {
  const { path, is404 = false } = route

  if (!is404 && !path) {
    throw new Error(`No path defined for route: ${JSON.stringify(route)}`)
  }
}

const consoleWarningForRouteWithoutNoIndex = route =>
  route.noIndex &&
  console.warn(`=> Warning: Route ${route.path} is using 'noIndex'. Did you mean 'noindex'?`)

export const createNormalizedRoute = (route, parent = {}, config = {}) => {
  const { children, ...routeWithOutChildren } = route
  const { path: parentPath = '/' } = parent
  const { tree: keepRouteChildren = false } = config
  const { is404 = false } = route

  throwErrorIfRouteIsMissingPath(route)

  const originalRoutePath = is404 ? PATH_404 : pathJoin(route.path)
  const routePath = is404 ? PATH_404 : pathJoin(parentPath, route.path)

  consoleWarningForRouteWithoutNoIndex(route)

  const normalizedRoute = {
    ...(keepRouteChildren ? route : routeWithOutChildren),
    path: routePath,
    originalPath: originalRoutePath,
    noindex: route.noindex || parent.noindex || route.noIndex,
    hasGetProps: !!route.getData,
  }

  return normalizedRoute
}

export const consoleWarningForMutlipleRoutesWithTheSamePath = routes => {
  routes.forEach(route => {
    const found = routes.filter(r => r.path === route.path)
    if (found.length > 1) {
      console.warn('More than one route is defined for path:', route.path)
    }
  })
}

const recurseCreateNormalizedRoute = (routes = [], parent, config) => {
  if (config.tree) {
    return routes
      .map(route => {
        const normalizedRoute = createNormalizedRoute(route, parent, config)
        normalizedRoute.children = recurseCreateNormalizedRoute(
          normalizedRoute.children,
          normalizedRoute,
          config
        )
        return normalizedRoute
      })
      .filter(d => d !== null)
  }

  return routes.reduce((memo = [], route) => {
    const normalizedRoute = createNormalizedRoute(route, parent, config)
    const routeExists = memo.find(({ path }) => path === normalizedRoute.path)

    return [
      ...memo,
      ...recurseCreateNormalizedRoute(route.children, normalizedRoute, config),
      ...(routeExists ? [] : [normalizedRoute]),
    ]
  }, [])
}

// Normalize routes with parents, full paths, context, etc.
export const normalizeRoutes = (routes, config = {}) => {
  const { disableDuplicateRoutesWarning, force404 = true } = config
  const normalizedRoutes = recurseCreateNormalizedRoute(routes, {}, config)

  if (!disableDuplicateRoutesWarning) {
    consoleWarningForMutlipleRoutesWithTheSamePath(normalizedRoutes)
  }

  if (force404 && !normalizedRoutes.find(r => r.is404)) {
    normalizedRoutes.push(
      createNormalizedRoute({
        is404: true,
        path: PATH_404,
      }),
      config
    )
  }

  return normalizedRoutes
}

// At least ensure the index page is defined for export
export const makeGetRoutes = config => async (...args) => {
  const { getRoutes = async () => DEFAULT_ROUTES } = config
  const routes = await getRoutes(...args)
  return normalizeRoutes(routes, config)
}

export const buildConfigation = (config = {}) => {
  // path defaults
  config.paths = {
    root: path.resolve(process.cwd()),
    src: 'src',
    dist: 'dist',
    devDist: 'tmp/dev-server',
    public: 'public',
    ...(config.paths || {}),
  }

  // Use the root to resolve all other relative paths
  const resolvePath = relativePath => path.resolve(config.paths.root, relativePath)

  // Resolve all paths
  const distPath =
    process.env.REACT_STATIC_ENV === 'development'
      ? resolvePath(config.paths.devDist || config.paths.dist)
      : resolvePath(config.paths.dist)

  const paths = {
    ROOT: config.paths.root,
    LOCAL_NODE_MODULES: path.resolve(getDirname(), '../../node_modules'),
    SRC: resolvePath(config.paths.src),
    DIST: distPath,
    PUBLIC: resolvePath(config.paths.public),
    NODE_MODULES: resolvePath('node_modules'),
    PACKAGE: resolvePath('package.json'),
    HTML_TEMPLATE: path.join(distPath, 'index.html'),
    STATIC_DATA: path.join(distPath, 'staticData'),
  }

  // Defaults
  const finalConfig = {
    // Defaults
    entry: path.join(paths.SRC, DEFAULT_ENTRY),
    getSiteData: () => ({}),
    renderToHtml: (render, Comp) => render(<Comp />),
    prefetchRate: 3,
    disableRouteInfoWarning: false,
    disableRoutePrefixing: false,
    outputFileRate: 10,
    // Config Overrides
    ...config,
    // Materialized Overrides
    paths,
    siteRoot: cutPathToRoot(config.siteRoot, '$1'),
    stagingSiteRoot: cutPathToRoot(config.stagingSiteRoot, '$1'),
    basePath: trimLeadingAndTrailingSlashes(config.basePath),
    stagingBasePath: trimLeadingAndTrailingSlashes(config.stagingBasePath),
    devBasePath: trimLeadingAndTrailingSlashes(config.devBasePath),
    extractCssChunks: config.extractCssChunks || false,
    inlineCss: config.inlineCss || false,
    getRoutes: makeGetRoutes(config),
  }

  // Set env variables to be used client side
  process.env.REACT_STATIC_PREFETCH_RATE = finalConfig.prefetchRate
  process.env.REACT_STATIC_DISABLE_ROUTE_INFO_WARNING = finalConfig.disableRouteInfoWarning
  process.env.REACT_STATIC_DISABLE_ROUTE_PREFIXING = finalConfig.disableRoutePrefixing

  return finalConfig
}

const buildConfigFromPath = configPath => {
  const filename = path.resolve(configPath)
  delete require.cache[filename]
  try {
    const config = require(configPath).default
    return buildConfigation(config)
  } catch (err) {
    console.error(err)
    return {}
  }
}

const fromFile = (configPath = DEFAULT_PATH_FOR_STATIC_CONFIG, watch) => {
  const config = buildConfigFromPath(configPath)

  if (watch) {
    chokidar.watch(configPath).on('all', () => {
      Object.assign(config, buildConfigFromPath(configPath))
      reloadRoutes()
    })
  }

  return config
}

// Retrieves the static.config.js from the current project directory
export default function getConfig (customConfig, { watch } = {}) {
  if (typeof customConfig === 'object') {
    // return a custom config obj
    return buildConfigation(customConfig)
  }

  // return a custom config file location
  // defaults to constant paath for static config
  return fromFile(customConfig, watch)
}
