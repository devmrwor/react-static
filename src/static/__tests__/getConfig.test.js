import getConfig, {
  cutPathToRoot,
  trimLeadingAndTrailingSlashes,
  createNormalizedRoute,
  makeGetRoutes,
  buildConfigation,
} from '../getConfig'
import defaultConfigDevelopment from '../__mocks__/defaultConfigDevelopment.mock'
import defaultConfigProduction from '../__mocks__/defaultConfigProduction.mock'

jest.mock('path', () => ({
  resolve: (stringOne = '/', stringTwo = '') => `${stringOne}${stringTwo}`,
  join: (stringOne, stringTwo) => `${stringOne}/${stringTwo}`,
  dirname: () => 'root/',
}))

jest.mock('../../utils/getDirname', () => () => './dirname/')

const testConfiguration = (configuration, configurationMock) => {
  expect(configuration).toMatchObject(configurationMock)
  expect(configuration.getSiteData).toBeInstanceOf(Function)
  expect(configuration.renderToHtml).toBeInstanceOf(Function)
  expect(configuration.getRoutes).toBeInstanceOf(Function)
}

describe('cutPathToRoot', () => {
  it('should return a root of the path', () => {
    expect(cutPathToRoot('./root/path/to/')).toBe('./root')
  })
})

describe('trimLeadingAndTrailingSlashes', () => {
  it('should return a String with the leading and trailing slash trimmed', () => {
    expect(trimLeadingAndTrailingSlashes('/path/to/')).toBe('path/to')
  })
})

describe('createNormalizedRoute', () => {
  describe('when working route is provided', () => {
    it('should return a normalized route', async () => {
      const route = createNormalizedRoute({ path: '/path/' })

      expect(route).toEqual({
        hasGetProps: false,
        noindex: undefined,
        originalPath: 'path',
        path: 'path',
      })
    })

    describe('when noindex is true', () => {
      it('should return a normalized route with noindex as true', () => {
        const route = createNormalizedRoute({ path: '/path/', noindex: true })

        expect(route.noindex).toEqual(true)
      })
    })

    describe('when noIndex is true', () => {
      let spy

      beforeEach(() => {
        spy = jest.spyOn(console, 'warn').mockImplementation(() => {})
      })

      it('should return a normalized route with noindex as true', () => {
        const route = createNormalizedRoute({ path: '/path/', noIndex: true })

        expect(route.noindex).toEqual(true)
      })

      it('should warns the user to use noIndex', () => {
        createNormalizedRoute({ path: '/path/', noIndex: true })

        expect(spy).toHaveBeenCalled()
        expect(spy).toBeCalledWith(
          "=> Warning: Route /path/ is using 'noIndex'. Did you mean 'noindex'?"
        )
      })

      afterEach(() => {
        spy.mockRestore()
      })
    })

    describe('when path is not defined', () => {
      it('should throw an error', () => {
        const route = { component: '/no/path/', noIndex: true }

        expect(() => createNormalizedRoute(route)).toThrow(
          `No path defined for route: ${JSON.stringify(route)}`
        )
      })

      describe('when route is 404', () => {
        it('should not throw an error', () => {
          expect(() => createNormalizedRoute({ component: '/no/path/', is404: true })).not.toThrow()
        })
      })
    })

    describe('when parent route is provided', () => {
      it('should return a normalized route', () => {
        const route = createNormalizedRoute({ path: '/to/' }, { path: '/path/' })

        expect(route).toEqual({
          hasGetProps: false,
          noindex: undefined,
          originalPath: 'to',
          path: 'path/to',
        })
      })

      describe('when parent noindex is true', () => {
        it('should return a normalized route with noindex as true', () => {
          const route = createNormalizedRoute({ path: '/to/' }, { path: '/path/', noindex: true })

          expect(route.noindex).toEqual(true)
        })
      })
    })
  })
})

describe('makeGetRoutes', () => {
  describe('when getRoutes is defined on config', () => {
    it('should return routes', async () => {
      const config = { getRoutes: async () => [{ path: '/path' }] }

      const getRoutes = makeGetRoutes(config)
      const routes = await getRoutes()

      expect(routes).toEqual([
        {
          hasGetProps: false,
          noindex: undefined,
          originalPath: 'path',
          path: 'path',
        },
      ])
    })

    it('should return routes', async () => {
      const config = {
        getRoutes: async () => [{ path: '/path' }, { is404: true, path: '404' }],
      }

      const getRoutes = makeGetRoutes(config)
      const routes = await getRoutes()

      expect(routes).toEqual([
        {
          hasGetProps: false,
          noindex: undefined,
          originalPath: 'path',
          path: 'path',
        },
        {
          hasGetProps: false,
          is404: true,
          noindex: undefined,
          originalPath: '404',
          path: '404',
        },
      ])
    })

    describe('when routes has children', () => {
      const routesWithChildren = [
        {
          path: '/path',
          children: [
            {
              path: 'to',
              children: [
                {
                  path: 'blog',
                },
                {
                  path: 'slug',
                },
              ],
            },
          ],
        },
      ]

      it('should return a flat Array of routes', async () => {
        const config = { getRoutes: async () => routesWithChildren }

        const getRoutes = makeGetRoutes(config)
        const routes = await getRoutes()

        expect(routes).toMatchSnapshot()
      })

      describe('when config.tree is defined', () => {
        it('should return a flat Array of routes', async () => {
          const config = {
            getRoutes: async () => routesWithChildren,
            tree: true,
          }

          const getRoutes = makeGetRoutes(config)
          const routes = await getRoutes()

          expect(routes).toMatchSnapshot()
        })
      })
    })
  })

  describe('when getRoutes is not defined on config', () => {
    it('should return default route', async () => {
      const config = {}

      const getRoutes = makeGetRoutes(config)
      const routes = await getRoutes()

      expect(routes).toEqual([
        {
          hasGetProps: false,
          noindex: undefined,
          originalPath: '/',
          path: '/',
        },
      ])
    })
  })
})

describe('buildConfigation', () => {
  let reactStaticEnviroment
  let reactStaticPrefetchRate
  let reactStaticDisableRouteInfoWarning
  let reactStaticDisableRoutePreFixing
  let spyProcess

  beforeEach(() => {
    reactStaticEnviroment = process.env.REACT_STATIC_ENV
    reactStaticPrefetchRate = process.env.REACT_STATIC_PREFETCH_RATE
    reactStaticDisableRouteInfoWarning = process.env.REACT_STATIC_DISABLE_ROUTE_INFO_WARNING
    reactStaticDisableRoutePreFixing = process.env.REACT_STATIC_DISABLE_ROUTE_PREFIXING
    spyProcess = jest.spyOn(process, 'cwd').mockImplementation(() => './root/')
  })

  describe('default configuration', () => {
    test('when REACT_STATIC_ENV is development', () => {
      process.env.REACT_STATIC_ENV = 'development'

      const configuration = buildConfigation()

      testConfiguration(configuration, defaultConfigDevelopment)
    })

    it('when REACT_STATIC_ENV is production', () => {
      process.env.REACT_STATIC_ENV = 'production'

      const configuration = buildConfigation()

      testConfiguration(configuration, defaultConfigProduction)
    })
  })

  test('REACT_STATIC_PREFETCH_RATE is set by the prefetchRate (default)', () => {
    process.env.REACT_STATIC_PREFETCH_RATE = null

    buildConfigation()

    expect(process.env.REACT_STATIC_PREFETCH_RATE).toBe('3')
  })

  test('REACT_STATIC_PREFETCH_RATE is set by the prefetchRate (from config)', () => {
    process.env.REACT_STATIC_PREFETCH_RATE = null

    buildConfigation({ prefetchRate: 10 })

    expect(process.env.REACT_STATIC_PREFETCH_RATE).toBe('10')
  })

  test('REACT_STATIC_DISABLE_ROUTE_INFO_WARNING is set by the disableRouteInfoWarning (default)', () => {
    process.env.REACT_STATIC_DISABLE_ROUTE_INFO_WARNING = null

    buildConfigation()

    expect(process.env.REACT_STATIC_DISABLE_ROUTE_INFO_WARNING).toBe('false')
  })

  test('REACT_STATIC_DISABLE_ROUTE_INFO_WARNING is set by the disableRouteInfoWarning (from config)', () => {
    process.env.REACT_STATIC_DISABLE_ROUTE_INFO_WARNING = null

    buildConfigation({ disableRouteInfoWarning: true })

    expect(process.env.REACT_STATIC_DISABLE_ROUTE_INFO_WARNING).toBe('true')
  })

  test('REACT_STATIC_DISABLE_ROUTE_PREFIXING is set by the disableRouteInfoWarning (default)', () => {
    process.env.REACT_STATIC_DISABLE_ROUTE_PREFIXING = null

    buildConfigation()

    expect(process.env.REACT_STATIC_DISABLE_ROUTE_PREFIXING).toBe('false')
  })

  test('REACT_STATIC_DISABLE_ROUTE_PREFIXING is set by the disableRouteInfoWarning (from config)', () => {
    process.env.REACT_STATIC_DISABLE_ROUTE_PREFIXING = null

    buildConfigation({ disableRoutePrefixing: true })

    expect(process.env.REACT_STATIC_DISABLE_ROUTE_PREFIXING).toBe('true')
  })

  afterEach(() => {
    process.env.REACT_STATIC_ENV = reactStaticEnviroment
    process.env.REACT_STATIC_PREFETCH_RATE = reactStaticPrefetchRate
    process.env.REACT_STATIC_DISABLE_ROUTE_INFO_WARNING = reactStaticDisableRouteInfoWarning
    process.env.REACT_STATIC_DISABLE_ROUTE_PREFIXING = reactStaticDisableRoutePreFixing

    spyProcess.mockRestore()
  })
})

describe('getConfig', () => {
  let spyProcess

  beforeEach(() => {
    spyProcess = jest.spyOn(process, 'cwd').mockImplementation(() => './root/')
  })

  describe('when no path or configuration is not provided', () => {
    it('should return a configuration using default file', () => {
      // mapped by the moduleNameMapper in package.js -> src/static/__mocks__/static.config.js
      // default path is 'static.config.js'
      const configuration = getConfig()

      testConfiguration(configuration, {
        ...defaultConfigProduction,
        entry: 'path/to/entry/index.js',
      })
    })
  })

  describe('when provided a path to configuration', () => {
    it('should return a configuration using file provided', () => {
      // mapped by the moduleNameMapper in package.js -> src/static/__mocks__/static.config.js
      const configuration = getConfig('./path/to/static.config.js')

      testConfiguration(configuration, {
        ...defaultConfigProduction,
        entry: 'path/to/entry/index.js',
      })
    })
  })

  describe('when provided a configuration', () => {
    it('should return a merged configuration', () => {
      const configuration = getConfig({
        entry: 'another/path/to/entry/index.js',
      })

      testConfiguration(configuration, {
        ...defaultConfigProduction,
        entry: 'another/path/to/entry/index.js',
      })
    })
  })

  afterEach(() => {
    spyProcess.mockRestore()
  })
})
