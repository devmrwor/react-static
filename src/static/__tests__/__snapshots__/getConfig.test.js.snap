// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeGetRoutes when getRoutes is defined on config when routes has children should return a flat Array of routes 1`] = `
Array [
  Object {
    "hasGetProps": false,
    "noindex": undefined,
    "originalPath": "path",
    "path": "path",
  },
  Object {
    "hasGetProps": false,
    "noindex": undefined,
    "originalPath": "to",
    "path": "path/to",
  },
  Object {
    "hasGetProps": false,
    "noindex": undefined,
    "originalPath": "blog",
    "path": "path/to/blog",
  },
  Object {
    "hasGetProps": false,
    "noindex": undefined,
    "originalPath": "slug",
    "path": "path/to/slug",
  },
]
`;

exports[`makeGetRoutes when getRoutes is defined on config when routes has children when config.tree is defined should return a flat Array of routes 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "hasGetProps": false,
            "noindex": undefined,
            "originalPath": "blog",
            "path": "path/to/blog",
          },
          Object {
            "children": Array [],
            "hasGetProps": false,
            "noindex": undefined,
            "originalPath": "slug",
            "path": "path/to/slug",
          },
        ],
        "hasGetProps": false,
        "noindex": undefined,
        "originalPath": "to",
        "path": "path/to",
      },
    ],
    "hasGetProps": false,
    "noindex": undefined,
    "originalPath": "path",
    "path": "path",
  },
]
`;
